### PR TITLE
Deactivate open offer if trigger price is reached

### PR DESCRIPTION
**Pull request by chimp1984**
Implements https://github.com/bisq-network/bisq/issues/4998

Based on https://github.com/bisq-network/bisq/pull/4993

### CODE REVIEW

File: desktop/src/main/java/bisq/desktop/main/PriceUtil.java [hunk 1]
```diff
@@ -0,0 +1,151 @@
+/*
```
Comment (ghubstan) [inline desktop/src/main/java/bisq/desktop/main/PriceUtil.java:1]: I instinctively need to ask if  `PriceUtil` could be moved to core.  I see dependencies on ChatView.log and ui validators, but much of the util logic could be made available to the api.

**Review by ghubstan — COMMENTED** [review]
I do see the offer disabling trigger price in the edit offer view, [EDIT] and **can** edit it.

But I can't see the disabling trigger price in the published offer views (below).  Squeezing it into the list view might not be feasible, but adding it to the other two might be trivial.

Confirm Offer View..
where-is-trigger-price

Published Offer List View...
where-is-trigger-price-in-offer-list

Published Offer Detail View...
where-is-trigger-price-in-published-offer-detail

File: desktop/src/main/java/bisq/desktop/main/PriceUtil.java (hunk 1 above)
Comment (chimp1984) [reply desktop/src/main/java/bisq/desktop/main/PriceUtil.java:1]: Yes I was thinking the same... was just to lazy to separate pure domain code with formatting stuff which is using desktop scope classes... But yes we should refactor that.

**Comment by chimp1984** [comment]
I was in the process to add a field there but as it only gets the offer passed but not the openOffer I was not sure... But probably its better to add it as well.
The triggerPrice is a field in openOffer not in Offer. Offer will be a part of trade and I think trigger price is a local setting of the user not a part of the mutual data which will be shared with the peer.

There is a trigger price column in openOffersView but it is hidden at default stage width as there is not enough space. If you scale larger it will get added. Similar we use in closed trades, there are a few detail columns only shown when width is larger. I am aware that many users might miss those hidden features, but have not found a good solution yet for it...
Maybe some icon on the table indicating that the table changes on resize, but as that is not a common UI feature I fear the icon will be just confusing...

**Comment by chimp1984** [comment]
Also note that I changed/fixed the above/below term in the price field description. Before it was only changing for BUY/SELL but not for fiat/altcoin. But as fiat price and altcoin price have inverted logic I applied that there as well.
E.g. For a BTC seller a higher fiat price is good but a lower altcoin price.

**Comment by ghubstan** [comment]
It's good that I can edit the trigger price... 
I am (reg) testing, so far so good.   
I wonder how difficult it would be to automatically re-enable an offer after the price goes back into the "enable range".
By the way, we shouldn't be doing this today ;-/

**Comment by chimp1984** [comment]
Screen Shot 2020-12-25 at 00 33 23
Screen Shot 2020-12-25 at 00 09 10
Screen Shot 2020-12-25 at 00 34 08

**Comment by chimp1984** [comment]
> I wonder how difficult it would be to automatically re-enable an offer after the price goes back into the "enable range".

Yes was also wondering if that might be good... Not sure also if we should show a popup once a offer gets deactivated. In th list view the shild icon gets yellow in that case. If price is still outside of permitted range the offer also cannot be re-activated directly (would trigger deactivation at next price update) but the user need to edit the offer... Not 100% sure if that is clear to users...

Also we do support that feature only for % based price, if you use fixed price you cannot set it as well if there is no price feed like BSQ. Wondering if it would make sense to allow it for those as well?

> By the way, we shouldn't be doing this today ;-/

Haha... I don't care much about xmas and holidays...

**Comment by chimp1984** [comment]
If offer got deactivated user need to edit offer and see the validation error:
Screen Shot 2020-12-25 at 12 36 40

**Comment by ghubstan** [comment]
> Haha... I don't care much about xmas and holidays...

We need time off just like religious people ;-)

> Not sure also if we should show a popup once a offer gets deactivated.

Not having one is OK for me.

> Wondering if it would make sense to allow it for those as well?

Yes, but maybe put that in another PR?

**Comment by ghubstan** [comment]
> the shield icon gets yellow in that case..

Ah, I see.  Didn't know what that icon meant but the tooltip explains.   Good, better that taking up more real-estate with a 'Trigger Price' column header.  I think it's fine the way it is.

File: core/src/main/java/bisq/core/offer/PriceEventHandler.java [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+        if (triggerPrice > 0) {
+            OfferPayload.Direction direction = openOffer.getOffer().getDirection();
+            boolean triggered = direction == OfferPayload.Direction.BUY ?
+                    marketPriceAsLong > triggerPrice :
+                    marketPriceAsLong < triggerPrice;
+            if (triggered) {
+                log.error("Market price exceeded the trigger price of the open offer. " +
+                                "We deactivate the open offer with ID {}. Currency: {}; offer direction: {}; " +
+                                "Market price: {}; Upper price threshold : {}",
+                        openOffer.getOffer().getShortId(),
```
Comment (ghubstan) [inline core/src/main/java/bisq/core/offer/PriceEventHandler.java]: Minor detail... placement of colons are not consistent (the last one).  
And you sure you wanted to use semi-colons to delimit key value pairs?  Is that consistent with other error log statements?

Comment (chimp1984) [reply core/src/main/java/bisq/core/offer/PriceEventHandler.java]: I think we have no style rule for that...

**Review by ghubstan — APPROVED** [review]
Tested ACK

**Review by ripcurlx — APPROVED** [review]
utACK based on https://github.com/bisq-network/bisq/pull/5001#pullrequestreview-559171289

**Pull request merged**